library/src/amd_detail/hipsolver_sparse.cpp

```diff
@@ -1074,4 +1086,299 @@ catch(...)
+                                       double                    tolerance,
+                                       int                       reorder,
+                                       float*                    x,
+                                       int*                      singularity)
+try
+{
+    if(!handle)
+        return HIPSOLVER_STATUS_NOT_INITIALIZED;
+    if(n < 0 || nnz < 0)
+        return HIPSOLVER_STATUS_INVALID_VALUE;
+    if(!descrA)
+        return HIPSOLVER_STATUS_INVALID_VALUE;
+    if(!csrRowPtr || !csrColInd || !csrVal || !descrA)
+        return HIPSOLVER_STATUS_INVALID_VALUE;
+    if(!b || !x || !singularity)
+        return HIPSOLVER_STATUS_INVALID_VALUE;
+    if(reorder < 0 || reorder > 3)
+        return HIPSOLVER_STATUS_INVALID_VALUE;
+
+    rocsparse_matrix_type mattype = rocsparse_get_mat_type((rocsparse_mat_descr)descrA);
+    rocsparse_index_base  indbase = rocsparse_get_mat_index_base((rocsparse_mat_descr)descrA);
+    if(mattype != rocsparse_matrix_type_general)
+        return HIPSOLVER_STATUS_MATRIX_TYPE_NOT_SUPPORTED;
+    if(indbase != rocsparse_index_base_zero && indbase != rocsparse_index_base_one)
+        return HIPSOLVER_STATUS_MATRIX_TYPE_NOT_SUPPORTED;
+
+    hipsolverSpHandle* sp = (hipsolverSpHandle*)handle;
+    *singularity          = -1;
+
+    // set up B
+    CHECK_HIP_ERROR(hipMemcpy((void*)x, b, sizeof(float) * n, hipMemcpyDeviceToDevice));
+
+    // convert A to dense matrix
+    float* denseA;
+    CHECK_HIP_ERROR(hipMalloc(&denseA, sizeof(float) * n * n));
+    rocsparse_scsr2dense(
+        sp->sphandle, n, n, (rocsparse_mat_descr)descrA, csrVal, csrRowPtr, csrColInd, denseA, n);
+
+    int* info;
+    CHECK_HIP_ERROR(hipMalloc(&info, sizeof(int)));
+
+    rocblas_status st
+        = rocsolver_sgels(sp->handle, rocblas_operation_none, n, n, 1, denseA, n, x, n, info);
+
+    // finalize singularity
+    CHECK_HIP_ERROR(hipMemcpy((void*)singularity, info, sizeof(int), hipMemcpyDeviceToHost));
+    *singularity = *singularity - 1;
+
+    CHECK_HIP_ERROR(hipFree(denseA));
+    CHECK_HIP_ERROR(hipFree(info));
+
+    return hipsolver::rocblas2hip_status(st);
+}
+catch(...)
+{
+    return hipsolver::exception2hip_status();
+}
+
+hipsolverStatus_t hipsolverSpDcsrlsvqr(hipsolverSpHandle_t       handle,
+                                       int                       n,
+                                       int                       nnz,
+                                       const hipsparseMatDescr_t descrA,
+                                       const double*             csrVal,
+                                       const int*                csrRowPtr,
+                                       const int*                csrColInd,
+                                       const double*             b,
+                                       double                    tolerance,
+                                       int                       reorder,
```

Comment:
`tolerance` and `reorder` are ignored because we solve with a dense matrix (with gels). 
The exact form of the QR factors doesn't matter when the output is only the solution vector X, however, a non-zero tolerance could give different singularity results for rocm and cuda for the same input matrix. 

This should be at least mentioned in the documentation (Maybe in the "some considerations when using SP API" section)

We could also have an internal `rocsolver_gels` for use by hipsolver that checks singularity within a given tolerance, but it is maybe too much if this is not going to be the final csrlsvqr implementation